### PR TITLE
Construct proper IOperation node for a RaiseEvent statement for a static custom event.

### LINKED ISSUE #23282: Unexpected IOperation tree for recursive RaiseEvent
```
        <CompilerTrait(CompilerFeature.IOperation)>
        <Fact>
        Public Sub RaiseEventRecursive()
            Dim source = <![CDATA[
Module M1
    Public Delegate Sub test_x()
    Public Custom Event t As test_x
        AddHandler(ByVal value As test_x)
        End AddHandler
        RemoveHandler(ByVal value As test_x)
        End RemoveHandler
        RaiseEvent()
            RaiseEvent t() 'BIND:"RaiseEvent t()"
        End RaiseEvent
    End Event
End Module
]]>.Value

            Dim expectedOperationTree = <![CDATA[
IInvalidOperation (OperationKind.Invalid, Type: null) (Syntax: 'RaiseEvent t()')
  Children(1):
      IInvocationOperation (Sub M1.raise_t()) (OperationKind.Invocation, Type: System.Void, IsImplicit) (Syntax: 'RaiseEvent t()')
        Instance Receiver: 
          null
        Arguments(0)
]]>.Value

            Dim expectedDiagnostics = <![CDATA[
BC41998: Statement recursively calls the containing 'RaiseEvent' for event 't'.
            RaiseEvent t() 'BIND:"RaiseEvent t()"
            ~~~~~~~~~~~~~~
]]>.Value

            VerifyOperationTreeAndDiagnosticsForTest(Of RaiseEventStatementSyntax)(source, expectedOperationTree, expectedDiagnostics)
        End Sub
```

Expected: IRaiseEventOperation rather than IInvalidOperation node
Microsoft.CodeAnalysis.Operations.VisualBasicOperationFactory.CreateBoundRaiseEventStatementOperation(BoundRaiseEventStatement boundRaiseEventStatement) shouldn't assert in debug build.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #23282.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review IOperation fix.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 1]
```diff
@@ -1377,38 +1377,36 @@ Namespace Microsoft.CodeAnalysis.Operations
...
-                ' This is for a custom event
-                boundInstance = receiver
+                receiverOpt = eventFieldAccess.ReceiverOpt
             End If
 
-            Dim eventReferenceInstance As Lazy(Of IOperation) = CreateReceiverOperation(boundInstance, eventSymbol)
+            Dim eventReferenceInstance As Lazy(Of IOperation) = CreateReceiverOperation(receiverOpt, eventSymbol)
 
             Dim eventReference As Lazy(Of IEventReferenceOperation) = New Lazy(Of IEventReferenceOperation)(Function() As IEventReferenceOperation
                                                                                                                 Return New LazyEventReferenceExpression(eventSymbol,
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Is this value of the Lazy(Of IEventReferenceOperation) also a Lazy(Of IEventReferenceOperation)? If so, is the extra laziness required?

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > Is this value of the Lazy(Of IEventReferenceOperation) also a Lazy(Of IEventReferenceOperation)?

Could you clarify the question please?

Comment (cston) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Never mind, I misread the lambda.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review IOperation fix.

**Review by jcouv — APPROVED** [review]
LGTM

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 2]
```diff
@@ -1377,38 +1377,36 @@ Namespace Microsoft.CodeAnalysis.Operations
...
                 Return New LazyInvalidOperation(children, _semanticModel, syntax, type, constantValue, isImplicit)
             End If
 
-            Dim receiver = eventInvocation.ReceiverOpt
-            Dim eventSymbol = boundRaiseEventStatement.EventSymbol
-            Dim eventReferenceSyntax = receiver.Syntax
+            Dim receiverOpt = eventInvocation.ReceiverOpt
+            Dim eventReferenceSyntax = If(receiverOpt?.Syntax,
+                                          If(TryCast(syntax, RaiseEventStatementSyntax)?.Name,
+                                             syntax))
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Is there a test hits this line of code? i.e. syntax is not a RaiseEventStatementSyntax.

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenEvents.vb [hunk 3]
```diff
@@ -426,6 +429,204 @@ End Module
...
+        OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+]]>.Value)
+
+            Assert.Equal("Event Program.E As Program.del1", model.GetSymbolInfo(raise.Name).Symbol.ToTestDisplayString())
+        End Sub
+
+        <Fact()>
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <WorkItem(23282, "https://github.com/dotnet/roslyn/issues/23282")>
+        Public Sub SimpleRaiseHandlerWithBlockEvent_02()
```
Comment (genlu) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenEvents.vb]: nit: instead of 01 and 02, mention `shared` and 'non-shared' in test method names.

**Review by genlu — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > Is there a test hits this line of code? i.e. syntax is not a RaiseEventStatementSyntax.

I don't think it is possible to get into this situation at the moment, but would like the code to be robust enough.

**Pull request merged**